include/swift/AST/ModuleDependencies.h

```diff
@@ -447,55 +460,176 @@ class ModuleDependenciesCache {
+
+  /// Record dependencies for the given module.
+  const ModuleDependencies *recordDependencies(StringRef moduleName,
+                                               ModuleDependencies dependencies);
+
+  /// Update stored dependencies for the given module.
+  const ModuleDependencies *updateDependencies(ModuleDependencyID moduleID,
+                                               ModuleDependencies dependencies);
+
+  /// Reference the list of all module dependencies.
+  const std::vector<ModuleDependencyID> &getAllModules() const {
+    return AllModules;
+  }
+};
+
+/// This "local" dependencies cache persists only for the duration of a given
+/// scanning action, and wraps an instance of a `GlobalModuleDependenciesCache`
+/// which may carry cached scanning information from prior scanning actions.
+/// This cache maintains a store of references to all dependencies found within
+/// the current scanning action (with their values stored in the global Cache),
+/// since these do not require clients to disambiguate them with search paths.
+class ModuleDependenciesCache {
+private:
+  GlobalModuleDependenciesCache &globalCache;
+
+  /// References to data in `globalCache` for dependencies accimulated during
+  /// the current scanning action.
+  std::unordered_map<ModuleDependenciesKind,
+                     llvm::StringMap<const ModuleDependencies *>,
+                     ModuleDependenciesKindHash>
+      ModuleDependenciesKindMap;
+
+  /// Retrieve the dependencies map that corresponds to the given dependency
+  /// kind.
+  llvm::StringMap<const ModuleDependencies *> &
+  getDependencyReferencesMap(ModuleDependenciesKind kind);
+  const llvm::StringMap<const ModuleDependencies *> &
+  getDependencyReferencesMap(ModuleDependenciesKind kind) const;
+
+  /// Local cache results lookup, only for modules which were discovered during
+  /// the current scanner invocation.
+  bool hasDependencies(StringRef moduleName,
+                       Optional<ModuleDependenciesKind> kind) const;
+
+  /// Local cache results lookup, only for modules which were discovered during
+  /// the current scanner invocation.
+  Optional<const ModuleDependencies *>
+  findDependencies(StringRef moduleName,
+                   Optional<ModuleDependenciesKind> kind) const;
+
+public:
+  ModuleDependenciesCache(GlobalModuleDependenciesCache &globalCache)
+      : globalCache(globalCache) {
+    ModuleDependenciesKindMap.insert(
+        {ModuleDependenciesKind::SwiftTextual,
+         llvm::StringMap<const ModuleDependencies *>()});
+    ModuleDependenciesKindMap.insert(
+        {ModuleDependenciesKind::SwiftBinary,
+         llvm::StringMap<const ModuleDependencies *>()});
+    ModuleDependenciesKindMap.insert(
+        {ModuleDependenciesKind::SwiftPlaceholder,
+         llvm::StringMap<const ModuleDependencies *>()});
+    ModuleDependenciesKindMap.insert(
+        {ModuleDependenciesKind::Clang,
+         llvm::StringMap<const ModuleDependencies *>()});
```

Comment:
I think these inserts could also be replaced by a loop over `FirstKind` to `LastKind`.